File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def check(now, word, cnt):
    ch = 0
    for w in range(cnt):
        if word[w] == now[w]:
            ch += 1
    if ch == (cnt-1):
        return True
    return False
    

def solution(begin, target, words):
    answer = 0
    
    cnt = len(begin)
    n = len(words)
    visit = [0]*n
    
    q = deque([(begin, 0)])
    
    while q:
        now = q.popleft()
        if now[0] == target:
            answer = now[1]
            break
        for i in range(n):
            if visit[i] == 0 and check(now[0], words[i], cnt):
                visit[i] = 1
                q.append((words[i], now[1]+1))
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    
    let cnt = begin.length;
    let n = words.length;
    let visit = Array.from(Array(n), ()=>(0));
    
    function check(now, word) {
        // now는 현재 단어, word는 비교할 단어
        // now와 word가 한 글자 다른 단어면 true, 아니면 false 반환
        let checkCnt = 0; // 같은 글자 세기
        for (let w=0; w<cnt; w++) {
            // 한글자씩 비교해서 한 글자 같으면 checkCnt += 1
            if (word[w] === now[w]) checkCnt++;
        }
        if (checkCnt === (cnt -1)) return true;
        
        return false;
    }
    
    q = [[begin, 0]];
    
    while (q.length > 0){
        let[now, dep] = q.shift();
        if (now === target) return dep;
        for (let i=0; i<n; i++){
            if (visit[i] === 0 && check(now, words[i])) {
                visit[i] = 1;
                q.push([words[i], dep+1]);
            }
        }   
    }
    
    return 0;
}